### PR TITLE
ability for efficient case conversion

### PR DESCRIPTION
**Pull request by hohwille**
* added `Case` enum with `UPPER_CASE` and `LOWER_CASE`.
* added `Junidecode.unidecode(String, Case)` method for efficient case conversion (much more efficient that calling Junidecode.unidecode(text).toLowerCase(Locale.ROOT)).
* extended JUnit test to cover new feature
* added feature to CHANGELOG

### CODE REVIEW

File: src/main/java/net/gcardone/junidecode/Junidecode.java [hunk 1]
```diff
@@ -627,9 +642,61 @@ public static String unidecode(final String s) {
              * one is reserved.
              */
             if (low < map.length) {
-                sb.append(map[low]);
+                caseConversion.append(sb, map[low]);
```
Comment (gcardone) [inline src/main/java/net/gcardone/junidecode/Junidecode.java:645]: I don't think that the conversion belongs on this line, because it causes a call to a function for each codepoint. It would be more efficient to do it once, for the whole string.

I suggest to implement this functionality inverting the dependency: `unidecode(final String s, final Case caseConversion)` should call `unidecode(final String s)` _and then_ apply the case conversion on the entire output. What do you think?

Comment (hohwille) [reply src/main/java/net/gcardone/junidecode/Junidecode.java:645]: I do not agree here as Java internally is designed very inefficient: A `String` is a wrapper for an array of characters (`char[]`). If you do `toLowerCase` on that, you have to create a new `char[]` and a new `String` wrapping it and again looping over every character to do the conversion. In that case, there would be no need at all to add this use-case to your library as then I can even just call `toLowerCase(Locale.ROOT)` you the result of junidecode. However, this is very inefficient if you process lots of text (especially since java does not know that we already have ASCII here and do not need to care about the complexity of case conversion with full unicode support).
If you care about the extra method call (that `JIT` will optimize away), I can remove the `ORIGINAL_CASE` and supply `null` instead so that using an `if (caseConversion==null)` the original code would still apply without the extra method call wrapping it.

Comment (hohwille) [reply src/main/java/net/gcardone/junidecode/Junidecode.java]: I updated the PR accordingly. Now it is guaranteed (even when running your code in GraalVM or TeaVM) that no extra Method call will happen if junidecode is used as before. The extra `if` will be optimized away by JIT as `caseConversion` is final and so if it once was `null` all further loop runs will produce the same result so no overhead for subsequent loopings. (If you would still be not convinced I would recommend to run some performance tests as otherwise, you would have to duplicate the code of the method and IMHO that would not be a good decision for maintenance reasons).

Comment (hohwille) [reply src/main/java/net/gcardone/junidecode/Junidecode.java]: No reaction? Do have to create my own fork of this project?

File: src/test/java/net/gcardone/junidecode/JunidecodeTest.java [hunk 2]
```diff
@@ -92,6 +96,11 @@ public void someTests() {
             String decoded = Junidecode.unidecode(sp.getLeft());
             assertEquals(sp.getRight(), decoded);
             sevenBitPurity(decoded);
+            // lower case
+            String lowerCase = Junidecode.unidecode(sp.getLeft(), Case.LOWER_CASE);
+            assertEquals(sp.getRight().toLowerCase(Locale.ROOT), lowerCase);
+            String upperCase = Junidecode.unidecode(sp.getLeft(), Case.UPPER_CASE);
+            assertEquals(sp.getRight().toUpperCase(Locale.ROOT), upperCase);
```
Comment (gcardone) [inline src/test/java/net/gcardone/junidecode/JunidecodeTest.java:104]: This is missing a test to ensure that `null` does not change the case.

File: src/main/java/net/gcardone/junidecode/Junidecode.java (hunk 1 above)
Comment (gcardone) [reply src/main/java/net/gcardone/junidecode/Junidecode.java]: Thanks, I much prefer the new version.

Sorry for the delay. On the bright side I refreshed my credentials with sonatype, so once your change is merged I can quickly release a new jar.

File: src/test/java/net/gcardone/junidecode/JunidecodeTest.java (hunk 2 above)
Comment (hohwille) [reply src/test/java/net/gcardone/junidecode/JunidecodeTest.java:104]: The existing test code above is calling `Junidecode.unidecode(sp.getLeft())` and testing the result.
And we know this:
```
  public static String unidecode(final String s) {

    return unidecode(s, null);
  }
```

So what are we missing to cover or what needs to be added?

Comment (hohwille) [reply src/test/java/net/gcardone/junidecode/JunidecodeTest.java:104]: FYI:
https://github.com/m-m-m/text/blob/master/ascii/src/main/java/io/github/mmm/text/ascii/impl/package-info.java

BTW: I found and fixed many many mistakes and one bug in Junitdecode. If you are interested I can give some details or your compare your implementation with mine.

Comment (gcardone) [reply src/test/java/net/gcardone/junidecode/JunidecodeTest.java:104]: > No reaction, no progress, no release. Simple PR now hanging for almost a year. Sorry to say so but I will create my own forked library of this code and solve the things I need on my own instead of being blocked so long...

Sorry, life takes over and I consider Junidecode fundamentally feature complete, so new features are low in my priority list. I should have stated this more openly.

To be clear: a fast path to convert ASCII-7 strings to uppercase and lowercase is more generic than Junidecode and there's no need for it to be coupled with `unidecode()`, so I think it's better implemented separately.

Looking at your library it has way bigger aspirations than Junidecode, so forking it seems to me a reasonable thing to do.

> I found and fixed many many mistakes and one bug in Junitdecode. If you are interested I can give some details or your compare your implementation with mine.

Please do open issues if you find bugs or missing codepoints (for example I'm aware that I'm missing the mapping from U+20BF (₿) to `BTC`).
